Upgrading account

tyrantdavis opened the pull request.

tyrantdavis committed:
Update views

tyrantdavis committed:
Implement timeoutable

tyrantdavis committed:
Install Stripe

tyrantdavis committed:
Install and configure stripe

tyrantdavis committed:
Generate charges controller

tyrantdavis committed:
Set premium user Charges controller

tyrantdavis committed:
Add Charges new view

tyrantdavis committed:
Add new action

tyrantdavis committed:
Edit Charges controller, update routes

tyrantdavis committed:
Update premium charges path, edit role views

tyrantdavis committed:
Generate downgrade controller,
Add user flow to downgrade_controller.rb

tyrantdavis committed:
Edited wiki_policy
Readjusted wikis displayed according to role

cbeck commented on app/policies/wiki_policy.rb:
```diff
@@ -1,3 +1,79 @@
+  end
+
+  def create?
+    user.admin? || user.standard? || user.premium?
```
If you are not logged in and this gets called, it will break. Since you are covering all three roles here, you prob can just go with user.present?

cbeck commented on app/policies/wiki_policy.rb:
```diff
@@ -1,3 +1,79 @@
+  end
+
+  def update?
+    user.present?
```
For this one, I might check to see if the user has the right to update. They'd need to be present, and either the record owner or admin (until you do collaborators, then one of those too).

cbeck commented on app/policies/wiki_policy.rb:
```diff
@@ -1,3 +1,79 @@
 class WikiPolicy < ApplicationPolicy
+  attr_reader :user, :wiki
 
+  def initialize(user, wiki)
```
You could have left this method out and just used and record everywhere. Also means you would get rid of the attr_reader on line 2

cbeck commented on app/policies/wiki_policy.rb:
```diff
@@ -1,3 +1,79 @@
+
+  def destroy?
+    # if the wiki belongs to the current user or user is admin
+    (wiki.user == user) || user.admin?
```
make sure to put user.present? as the first condition in this

cbeck commented on app/controllers/downgrade_controller.rb:
```diff
@@ -0,0 +1,17 @@
+  end
+
+  def create
+    current_user.update_attribute(:role, 0)
```
or current_user.standard!

cbeck commented on app/controllers/downgrade_controller.rb:
```diff
@@ -0,0 +1,17 @@
+
+  def create
+    current_user.update_attribute(:role, 0)
+    @wikis = Wiki.all
```
You can automatically scope this to the current_user like so:

```
@wikis = current_user.wikis
```

cbeck commented on app/controllers/downgrade_controller.rb:
```diff
@@ -0,0 +1,17 @@
+    current_user.update_attribute(:role, 0)
+    @wikis = Wiki.all
+    @wikis.each do |wiki|
+      if wiki.user.id == current_user.id
```
Then you wouldn't need this conditional

cbeck reviewed (changes requested):
Made a few suggestions. Not all of them are critical for approval, but there are a couple I'd like to see changed.

tyrantdavis committed:
Remove redundancies from wiki policy.
Update model object reference. Refactor query methods.

if v:progname =~? "evim"
  finish
endif

" Use Vim settings, rather than Vi settings (much better!).
" This must be first, because it changes other options as a side effect.
set nocompatible

" allow backspacing over everything in insert mode
set backspace=indent,eol,start

if has("vms")
  set nobackup		" do not keep a backup file, use versions instead
else
  set backup		" keep a backup file
endif
set history=50		" keep 50 lines of command line history
set ruler		" show the cursor position all the time
set showcmd		" display incomplete commands
set incsearch		" do incremental searching

" For Win32 GUI: remove 't' flag from 'guioptions': no tearoff menu entries
" let &guioptions = substitute(&guioptions, "t", "", "g")

" Don't use Ex mode, use Q for formatting
map Q gq

" CTRL-U in insert mode deletes a lot.  Use CTRL-G u to first break undo,
" so that you can undo CTRL-U after inserting a line break.
inoremap <C-U> <C-G>u<C-U>

" In many terminal emulators the mouse works just fine, thus enable it.
if has('mouse')
  set mouse=a
endif

" Switch syntax highlighting on, when the terminal has colors
" Also switch on highlighting the last used search pattern.
if &t_Co > 2 || has("gui_running")
  syntax on
  set hlsearch
endif

" Only do this part when compiled with support for autocommands.
if has("autocmd")

  " Enable file type detection.
  " Use the default filetype settings, so that mail gets 'tw' set to 72,
  " 'cindent' is on in C files, etc.

cbeck commented on app/controllers/downgrade_controller.rb:
```diff
@@ -1,17 +1,17 @@
   def create
-    current_user.update_attribute(:role, 0)
-    @wikis = Wiki.all
+    current_user.standard?
```
This isn't actually doing anything

cbeck commented on app/controllers/downgrade_controller.rb:
```diff
@@ -1,17 +1,17 @@
-      if wiki.user.id == current_user.id
+      # ... make private wikis to public.
         wiki.update_attribute(:private, false)
       end
```
Does this work? In this change, it looks like you have an extra end here that wasn't deleted when you got rid of your if statement

cbeck commented on app/policies/wiki_policy.rb:
```diff
@@ -1,21 +1,14 @@
   def update?
-    user.present?
+    user.present? && ((record.user == user) || user.admin?)
   end
```
Perfect!

cbeck reviewed (approved):
The changes look good! I did have a question about something I saw in the downgrade_controller on line 11 - looks like you have an extra end. If you test it and it breaks, that'd be why. If all is working, then cool! Must have been just a git diff thing that was throwing me off.

Thanks for making the changes!
